src/esockd_acceptor.erl

```diff
@@ -237,32 +270,64 @@ handle_accept_error(enotconn, _, _) ->
     ok;
 handle_accept_error(einval, _, _) ->
     ok;
-handle_accept_error(overloaded, _, #state{proto = Proto, listen_on = ListenOn}) ->
-    esockd_server:inc_stats({Proto, ListenOn}, closed_overloaded, 1),
+handle_accept_error(overloaded, _, _) ->
     ok;
 handle_accept_error(Reason, Msg, #state{sockname = Sockname}) ->
-    error_logger:error_msg(Msg, [esockd:format(Sockname), Reason]).
+    logger:log(error, #{msg => Msg,
+                        listener => esockd:format(Sockname),
+                        cause => Reason}).
 
-handle_socket_error(closed, State) ->
+handle_socket_error(closed, State, _StateName) ->
     {stop, normal, State};
 %% {error, econnaborted} -> accept
-%% {error, esslaccept}   -> accept
-%% {error, esslaccept}   -> accept
-handle_socket_error(Reason, State) when Reason =:= econnaborted; Reason =:= esslaccept ->
+handle_socket_error(Reason, State, suspending) when Reason =:= econnaborted ->
+    {keep_state, State, {next_event, internal, accept_and_close}};
+handle_socket_error(Reason, State, _StateName) when Reason =:= econnaborted ->
     {next_state, waiting, State, {next_event, internal, begin_waiting}};
 %% emfile: The per-process limit of open file descriptors has been reached.
 %% enfile: The system limit on the total number of open files has been reached.
-%% enfile: The system limit on the total number of open files has been reached.
-handle_socket_error(Reason, State) when Reason =:= emfile; Reason =:= enfile ->
-    error_logger:error_msg(
-        "Accept error on ~s: ~s",
-        [esockd:format(State#state.sockname), explain_posix(Reason)]
-    ),
-    {next_state, suspending, State, {state_timeout, 1000, begin_waiting}};
-handle_socket_error(Reason, State) ->
+handle_socket_error(Reason, State, suspending) when Reason =:= emfile; Reason =:= enfile ->
+    log_system_limit(State, Reason),
+    {keep_state, State, {next_event, internal, accept_and_close}};
+handle_socket_error(Reason, State, _StateName) when Reason =:= emfile; Reason =:= enfile ->
+    log_system_limit(State, Reason),
+    start_suspending(State, 1000);
+handle_socket_error(Reason, State, _StateName) ->
     {stop, Reason, State}.
 
 explain_posix(emfile) ->
     "EMFILE (Too many open files)";
 explain_posix(enfile) ->
     "ENFILE (File table overflow)".
+
+log_system_limit(State, Reason) ->
+    logger:log(critical,
+               #{msg => "cannot_accept_more_connections",
+                 listener => esockd:format(State#state.sockname),
+                 cause => explain_posix(Reason)}).
+
+start_suspending(State, Timeout) ->
+    Actions = [{next_event, internal, accept_and_close},
+               {state_timeout, Timeout, begin_waiting}],
+    {next_state, suspending, State, Actions}.
+
+inc_stats(#state{proto = Proto, listen_on = ListenOn}, Tag) ->
+    Counter = counter(Tag),
+    _ = esockd_server:inc_stats({Proto, ListenOn}, Counter, 1),
+    ok.
+
+counter(accepted) -> accepted;
+counter(emfile) -> closed_overloaded;
```

Comment:
```suggestion
counter(emfile) -> closed_sys_limit;
```
Prefer map to other such as `sys_limit`, becasue overloaded is reserved for runq overload.